cfme/physical/physical_rack.py

```diff
@@ -0,0 +1,336 @@
+
+from cfme.base.login import BaseLoggedInPage
+from cfme.common import Taggable
+from cfme.exceptions import (
+    ItemNotFound,
+    StatsDoNotMatch,
+    RackStatsDoesNotContain,
+    ProviderHasNoProperty
+)
+from cfme.modeling.base import BaseEntity, BaseCollection
+from cfme.utils.appliance.implementations.ui import CFMENavigateStep, navigate_to, navigator
+from cfme.utils.log import logger
+from cfme.utils.pretty import Pretty
+from cfme.utils.providers import get_crud_by_name
+from cfme.utils.update import Updateable
+from cfme.utils.wait import wait_for
+
+
+@attr.s
+class PhysicalRack(BaseEntity, Updateable, Pretty, Taggable):
+    """Model of an Physical Rack in cfme.
+
+    Args:
+        name: Name of the physical rack.
+        custom_ident: The custom identifiter.
+        provider: The physical provider.
+
+    Usage:
+        myrack = PhysicalRack()
+        myrack.create()
+
+    """
+    pretty_attrs = ['name', 'custom_ident']
+
+    name = attr.ib()
+    provider = attr.ib(default=None)
+    custom_ident = attr.ib(default=None)
+    db_id = None
+
+    INVENTORY_TO_MATCH = ['power_state']
+    STATS_TO_MATCH = ['cores_capacity', 'memory_capacity']
+
+    def load_details_page(self):
+        """Load the PhysicalRack details page."""
+        view = navigate_to(self, "Details", use_resetter=False, wait_for_view=True)
+        return view
+
+    def refresh(self, cancel=False):
+        """Perform 'Refresh Relationships and Power States' for the rack.
+
+        Args:
+            cancel (bool): Whether the action should be cancelled, default to False
+        """
+        view = navigate_to(self, "Details")
+        view.toolbar.configuration.item_select("Refresh Relationships and Power States",
+            handle_alert=cancel)
+
+    def wait_for_physical_rack_state_change(self, desired_state, timeout=300):
+        """Wait for PhysicalRack to come to desired state. This function waits just the needed amount of
+           time thanks to wait_for.
+
+        Args:
+            desired_state (str): 'on' or 'off'
+            timeout (int): Specify amount of time (in seconds) to wait until TimedOutError is raised
+        """
+        view = navigate_to(self.parent, "All")
+
+        def _looking_for_state_change():
+            entity = view.entities.get_entity(name=self.name)
+            return "currentstate-{}".format(desired_state) in entity.data['state']
+
+        wait_for(_looking_for_state_change, fail_func=view.browser.refresh, num_sec=timeout)
+
+    @property
+    def exists(self):
+        """Checks if the physical_rack exists in the UI.
+
+        Returns: :py:class:`bool`
+        """
+        view = navigate_to(self.parent, "All")
+        try:
+            view.entities.get_entity(name=self.name, surf_pages=True)
+        except ItemNotFound:
+            return False
+        else:
+            return True
+
+    @cached_property
+    def get_db_id(self):
```

Comment:
This is an odd pattern to combine with cached_property and prevents proper cache management.

There's no need for a db_id class attribute that can't be set during construction, then a cached_property with a `get_*` name. `get_db_id`, as a method name, does not sound like a property, but a bound method.

I think you should have a cached property called `db_id` that returns the result of `self.appliance.physical_rack_id(self.name)`

But this begs the question - if its a single line call to the appliance method, passing the rack name, why have this property at all when the caller could look up the ID from the appliance method directly?

Reply:
It makes sense, so I removed this method